new-log-viewer/src/contexts/UrlContextProvider.tsx

```diff
@@ -0,0 +1,254 @@
+import React, {
+    createContext,
+    useEffect,
+    useState,
+} from "react";
+
+import {NullableProperties} from "../typings/common";
+import {
+    HASH_PARAM_NAMES,
+    SEARCH_PARAM_NAMES,
+    UrlHashParams,
+    UrlHashParamUpdatesType,
+    UrlParamsType,
+    UrlSearchParams,
+    UrlSearchParamUpdatesType,
+} from "../typings/url";
+import {getAbsoluteUrl} from "../utils/url";
+
+
+const UrlContext = createContext <UrlParamsType>({} as UrlParamsType);
+
+/**
+ * Default values of the search parameters.
+ */
+const URL_SEARCH_PARAMS_DEFAULT = Object.freeze({
+    [SEARCH_PARAM_NAMES.FILE_PATH]: null,
+});
+
+/**
+ * Default values of the hash parameters.
+ */
+const URL_HASH_PARAMS_DEFAULT = Object.freeze({
+    [HASH_PARAM_NAMES.LOG_EVENT_NUM]: null,
+});
+
+/**
+ * Computes updated URL search parameters based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the search parameters. If a value
+ * is `null`, the corresponding kv-pair will be removed from the updated search parameters.
+ * @return The updated search parameters string.
+ */
+const getUpdatedSearchParams = (updates: UrlSearchParamUpdatesType) => {
+    const newSearchParams = new URLSearchParams(window.location.search.substring(1));
+    const {filePath: newFilePath} = updates;
+
+    for (const [key, value] of Object.entries(updates)) {
+        if (SEARCH_PARAM_NAMES.FILE_PATH as string === key) {
+            // Updates to `filePath` should be handled last.
+            continue;
+        }
+        if (null === value) {
+            newSearchParams.delete(key);
+        } else {
+            newSearchParams.set(key, String(value));
+        }
+    }
+
+    // `filePath` should always be the last search parameter
+    const originalFilePath = newSearchParams.get(SEARCH_PARAM_NAMES.FILE_PATH);
+    newSearchParams.delete(SEARCH_PARAM_NAMES.FILE_PATH);
+    if ("undefined" === typeof newFilePath && null !== originalFilePath) {
+        // If no change in `filePath` is specified, put the original `filePath` back.
+        newSearchParams.set(SEARCH_PARAM_NAMES.FILE_PATH, originalFilePath);
+    } else if ("undefined" !== typeof newFilePath && null !== newFilePath) {
+        // If the new `filePath` has a non-null value, set the value.
+        newSearchParams.set(SEARCH_PARAM_NAMES.FILE_PATH, newFilePath);
+    }
```

Comment:
```suggestion
    if ("undefined" !== typeof newFilePath && null !== newFilePath) {
        newSearchParams.set(SEARCH_PARAM_NAMES.FILE_PATH, newFilePath);
    } else if (null !== originalFilePath) {
        // No change in `filePath` is specified, so restore the original `filePath`.
        newSearchParams.set(SEARCH_PARAM_NAMES.FILE_PATH, originalFilePath);
    }
```

I think that's a little clearer?

Reply:
This might not work if `originalFilePath` is non-null but `newFilePath` is null (if the function user wishes to delete the parameter).

Reply:
True, my mistake.